azure/scope/cluster.go

```diff
@@ -880,7 +880,15 @@ func (s *ClusterScope) SetFailureDomain(id string, spec clusterv1.FailureDomainS
 	if s.AzureCluster.Status.FailureDomains == nil {
 		s.AzureCluster.Status.FailureDomains = make(clusterv1.FailureDomains)
 	}
-	s.AzureCluster.Status.FailureDomains[id] = spec
+
+	if len(s.AzureCluster.Spec.FailureDomains) == 0 {
+		s.AzureCluster.Status.FailureDomains[id] = spec
+		return
+	}
+
+	if fd, ok := s.AzureCluster.Spec.FailureDomains[id]; ok && fd.ControlPlane {
```

Comment:
actually, on second thought, there is a change to current flow, albeit a bit exotic. but let's definitely discuss if this should be changed or not

previously, when FD would be discovered (and not suitable for control plane, aka `controlPlane: false`), it would get into the status field no matter what

in current code, imagine there is FD1 discovered with `controlPlane: false` and in `spec`, there is FD2 with an arbitrary value

the change is, FD1 will no longer be announced in the status.

thoughts?

Reply:
> "FD is always propagated to status, only its controlPlane value can be overriden to false by spec"

This is more future-proof and safe, so +1

In practice, CAPZ sets every single Azure zone it discovers to `controlPlane: true`: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/controllers/azurecluster_reconciler.go#L217 so this shouldn't actually happen